Latest version can not run on Windows 7 64bit

Latest git version(2020-10-11) can not run on Windows 7 64 bit ( in VirtualBox ).

for example widget_gallery.exe:

INFO  [druid_shell::platform::windows::util] No shcore.dll
INFO  [druid_shell::platform::windows::util] Could not load `GetDpiForSystem`. Windows 10 or later is needed
INFO  [druid_shell::platform::windows::util] Could not load `GetDpiForWindow`. Windows 10 or later is needed
INFO  [druid_shell::platform::windows::util] Could not load `SetProcessDpiAwarenessContext`. Windows 10 or later is needed
INFO  [druid_shell::platform::windows::util] Could not load `GetSystemMetricsForDpi`. Windows 10 or later is needed
DEBUG [druid::localization] available locales [], current en-US
DEBUG [druid::localization] resolved: [en-US]
DEBUG [druid_shell::platform::windows::window] dxgi factory pointer = 0x3a39d0
DEBUG [druid_shell::platform::windows::window] 0x3b6bb0: desc = Some("VirtualBox Graphics Adapter (WDDM)"), vram = 147914752
DEBUG [druid_shell::platform::windows::window] adapter = 0x3b6bb0
ERROR [druid_shell::platform::windows::dcomp] D3D11CreateDevice: 0x887a0004
ERROR [druid_shell::platform::windows::window] Creating swapchain failed: Hr(-2005270524)
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', druid-shell\src\platform\windows\window.rs:812:55
stack backtrace:
   0:        0x13f93479e - <unknown>
   1:        0x13f94b57b - <unknown>
   2:        0x13f930b98 - <unknown>
   3:        0x13f937534 - <unknown>
   4:        0x13f937118 - <unknown>
   5:        0x13f937def - <unknown>
   6:        0x13f937955 - <unknown>
   7:        0x13f9350df - <unknown>
   8:        0x13f937909 - <unknown>
   9:        0x13f9492d0 - <unknown>
  10:        0x13f94921c - <unknown>
  11:        0x13f857d06 - <unknown>
  12:        0x13f85bf6d - <unknown>
  13:         0x77079bbd - TranslateMessageEx
  14:         0x7707729b - SetWindowTextW
  15:         0x770767e9 - IsDialogMessageW
  16:         0x771cb5cf - KiUserCallbackDispatcher
  17:         0x770718ea - ShowWindow
  18:        0x13f84e70a - <unknown>
  19:        0x13f73386c - <unknown>
  20:        0x13f749dd8 - <unknown>
  21:        0x13f739d76 - <unknown>
  22:        0x13f739d8c - <unknown>
  23:        0x13f938133 - <unknown>
  24:        0x13f74fd07 - <unknown>
  25:        0x13f9aefc4 - <unknown>
  26:         0x76f5556d - BaseThreadInitThunk
  27:         0x771b385d - RtlUserThreadStart

30922d0 can not compile here.
I mean 0.6 (2020-6-2) release version is ok. but git version(2020-10-11) is not ok.

30922d0 compile error message:
error[E0599]: no variant or associated item named `ForegroundColor` found for enum `TextAttribute` in the current scope
   --> druid\src\text\layout.rs:231:55
    |
231 |                     .default_attribute(TextAttribute::ForegroundColor(text_color))
    |                                                       ^^^^^^^^^^^^^^^ variant or associated item not found in `TextAttribute`

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -106,6 +106,7 @@ You can find its changes [documented below](#060---2020-06-01).
 - Able to select text in multiple TextBoxes at once. ([#1276] by [@sysint64])
 - The scroll bar now shows when the contents of a scrollable area change size. ([#1278] by [@Majora320])
 - Fix `widget::Either` using the wrong paint insets ([#1299] by [@andrewhickman])
+- Improve Windows 7 DXGI compatibility ([#1311] by [@raphlinus])
 
 ### Visual
 
@@ -499,6 +500,7 @@ Last release without a changelog :(
 [#1295]: https://github.com/linebender/druid/pull/1280
 [#1298]: https://github.com/linebender/druid/pull/1298
 [#1299]: https://github.com/linebender/druid/pull/1299
+[#1311]: https://github.com/linebender/druid/pull/1311
 
 [Unreleased]: https://github.com/linebender/druid/compare/v0.6.0...master
 [0.6.0]: https://github.com/linebender/druid/compare/v0.5.0...v0.6.0

--- a/druid-shell/src/platform/windows/dcomp.rs
+++ b/druid-shell/src/platform/windows/dcomp.rs
@@ -23,7 +23,7 @@ use log::error;
 
 use winapi::shared::winerror::SUCCEEDED;
 use winapi::um::d3d11::*;
-use winapi::um::d3dcommon::D3D_DRIVER_TYPE_HARDWARE;
+use winapi::um::d3dcommon::{D3D_DRIVER_TYPE_HARDWARE, D3D_DRIVER_TYPE_WARP};
 use winapi::um::winnt::HRESULT;
 use winapi::Interface;
 use wio::com::ComPtr;
@@ -45,21 +45,29 @@ pub struct D3D11Device(ComPtr<ID3D11Device>);
 impl D3D11Device {
     /// Creates a new device with basic defaults.
     pub(crate) fn new_simple() -> Result<D3D11Device, HRESULT> {
+        let mut hr = 0;
         unsafe {
             let mut d3d11_device: *mut ID3D11Device = null_mut();
-            let flags = D3D11_CREATE_DEVICE_BGRA_SUPPORT; // could probably set single threaded
-            let hr = D3D11CreateDevice(
-                null_mut(),
-                D3D_DRIVER_TYPE_HARDWARE,
-                null_mut(),
-                flags,
-                null(),
-                0,
-                D3D11_SDK_VERSION,
-                &mut d3d11_device,
-                null_mut(),
-                null_mut(),
-            );
+            // Note: could probably set single threaded in flags for small performance boost.
+            let flags = D3D11_CREATE_DEVICE_BGRA_SUPPORT;
+            // Prefer hardware but use warp if it's the only driver available.
+            for driver_type in &[D3D_DRIVER_TYPE_HARDWARE, D3D_DRIVER_TYPE_WARP] {
+                hr = D3D11CreateDevice(
+                    null_mut(),
+                    *driver_type,
+                    null_mut(),
+                    flags,
+                    null(),
+                    0,
+                    D3D11_SDK_VERSION,
+                    &mut d3d11_device,
+                    null_mut(),
+                    null_mut(),
+                );
+                if SUCCEEDED(hr) {
+                    break;
+                }
+            }
             if !SUCCEEDED(hr) {
                 error!("D3D11CreateDevice: 0x{:x}", hr);
             }

